tomography_preprocessing/utils/aretomo.py

```diff
@@ -0,0 +1,185 @@
+    idx_min = []
+    for angle in tlt_angles:
+        arr_diff = abs(star_angles - angle)
+        #If difference between nominal tilt and tlt file is less than 0.5, assume same tilt
+        if np.min(arr_diff) < 0.5:	
+            idx_min.append(np.argmin(arr_diff))
+    complete_df = tilt_image_df.copy()
+    complete_df = complete_df.iloc[idx_min]
+    return complete_df
+
+
+def write_relion_tilt_series_alignment_output(
+        tilt_image_df: pd.DataFrame,
+        tilt_series_id: str,
+        pixel_size: float,
+        imod_directory: Path,
+        output_star_file: Path,
+):
+    shifts_px, euler_angles = get_tilt_series_alignment_parameters(imod_directory, tilt_series_id)
+    tilt_image_df = remove_ignored_images(tilt_image_df, euler_angles)
+    tilt_image_df[['rlnOriginXAngst', 'rlnOriginYAngst']] = shifts_px * pixel_size
+    tilt_image_df[['rlnAngleRot', 'rlnAngleTilt', 'rlnAnglePsi']] = euler_angles
+
+    starfile.write({tilt_series_id: tilt_image_df}, output_star_file)
+
+
+def relion_tilt_series_alignment_parameters_to_relion_matrix(
+        shifts: pd.DataFrame,
+        euler_angles: pd.DataFrame,
+        tilt_image_dimensions: np.ndarray,
+        tomogram_dimensions: np.ndarray,
+):
+    """shifts need to be in px at this point"""
+    tilt_image_center = (tilt_image_dimensions - 1) / 2
+    n_tilt_images = shifts.shape[0]
+    shifts = np.c_[shifts, np.zeros(n_tilt_images)]  # zero-fill (n, 2) to (n, 3)
+    tomogram_center_to_origin = -tomogram_dimensions / 2
+    tilt_image_corner_to_center_vector = np.append(tilt_image_center, 0)
+
+    # Generate affine matrices for each component of the final transformation
+    # first the rotations
+    in_plane_rotation = Rz(euler_angles['rlnAnglePsi'])
+    tilt = Ry(euler_angles['rlnAngleTilt'])
+
+    # offsets
+    tomogram_origin_offset = S(tomogram_center_to_origin)
+    tilt_image_corner_to_center = S(tilt_image_corner_to_center_vector)
+    tilt_image_centering_shift = S(shifts)
+
+    # compose matrices
+    volume_transforms = in_plane_rotation @ tilt @ tomogram_origin_offset
+    image_transforms = tilt_image_corner_to_center @ tilt_image_centering_shift
+    return np.squeeze(image_transforms @ volume_transforms)
+
+
+def create_job_directory_structure(output_directory: Path) -> Tuple[Path, Path]:
+    """Create directory structure for an AreTomo alignment job."""
+    tilt_series_directory = output_directory / 'tilt_series'
+    tilt_series_directory.mkdir(parents=True, exist_ok=True)
+
+    imod_alignments_directory = output_directory / 'aretomo_alignments'
+    imod_alignments_directory.mkdir(parents=True, exist_ok=True)
+    return tilt_series_directory, imod_alignments_directory
+
+
+def align_single_tilt_series(
+        tilt_series_id: str,
+        tilt_series_df: pd.DataFrame,
+        tilt_image_df: pd.DataFrame,
+        alignment_function: Callable,
+        alignment_function_kwargs: Dict[str, Any],
+        output_directory: Path,
+):
```

Comment:
if this is identical to the IMOD structure please reuse the IMOD util, maybe refactoring it into a more general ts alignment util function

Reply:
calls a more elaborate alignment function inside, but yes a more general approach here may help